### PR TITLE
cask/download: fix `cleanup` so that it does prune old cask downloads

### PR DESCRIPTION
**Pull request by aholland**
## Problem

`brew cleanup` (and `brew cleanup -s`) does not prune a cask's superseded downloads when a newer version replaces them, for any cask whose download URL basename ≠ its token — the common case (e.g. `docker`, `microsoft-edge`, `firefox`, `discord`). Old versions accumulate until the generic `HOMEBREW_CLEANUP_MAX_AGE_DAYS` (default 120) sweep removes them, so a frequently-updated cask can hold many GB of stale installers. (Age-expired downloads *are* still removed — that part works; the bug is that superseded versions aren't removed when replaced.)

## Cause

Since `44f058edb5` (the formula/cask/Ruby-source download unification) a cask's cache symlink is named from `File.basename(url)` — the vendor's URL basename, e.g. `Docker.dmg--<version>` — instead of the cask token. `Cleanup#stale_cask?` identifies the cached cask by parsing the name prefix as a token (`CaskLoader.load(prefix)`); a vendor filename is not a loadable token, so it returns early, no version is ever marked stale, and the symlinks (and the blobs they keep referenced) are never removed.

Formulae are unaffected: Homebrew hosts their bottles, so the bottle basename already begins with the formula name. Casks whose URL basename is long enough are also unaffected, because `download_name` then falls back to `cask.full_token` (`b1bd71c412`) — which is why the breakage is intermittent across casks.

## Fix

Name the cask download symlink by the cask token (its pre-`44f058edb5` behaviour). Only the symlink changes: `AbstractFileDownloadStrategy#cached_location` names the cached file from the URL (`resolved_basename`), never from `download_name`, so the blob keeps its real `Docker.dmg` name. With the symlink token-named, `stale_cask?` resolves it again and `cleanup_cask`'s `Cask/#{cask.token}--*` glob matches. No cleanup logic changes.

No migration is needed: existing URL-named symlinks age out within the max-age window, so caches self-correct within ≤120 days.

### Notes

- Uses `cask.token`, not `full_token`: `stale_cask?` parses with a non-greedy `\A(.*?)--`, so a slash-substituted full token (`a--b--c--…`) would mis-parse; `token` is also what `cleanup_cask` globs.
- Removes the length cap added in `b1bd71c412`: a token is always short, and that cap's fallback already kept the (unbounded) version, so it never bounded the symlink length beyond swapping a long basename for the token.
- Test coverage: the existing `test/cask/download_spec.rb` `#download_name` unit spec is updated to assert the token (the previous examples asserted the URL-basename behaviour this PR changes).

## Steps to reproduce

1. `brew install --cask docker` (any cask whose URL basename ≠ its token).
2. Accumulate two or more downloaded versions over time.
3. `ls "$(brew --cache)/Cask/" | grep -i docker` → multiple `Docker.dmg--<version>.dmg` symlinks.
4. `brew cleanup -s`.
5. Before this change the old versions remain in `$(brew --cache)/downloads/`; after, only the current version remains (as already happens for formulae).

-----

- [x] Have you followed our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) guidelines?
- [x] Have you checked for other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you explained what your changes do? Performance claims (e.g. "this is faster") must include [Hyperfine](https://github.com/sharkdp/hyperfine) benchmarks.
- [x] Have you explained why you'd like these changes included, not just what they do?
- [x] For bug fixes, have you given step-by-step `brew` commands to reproduce the bug?
- [x] Have you written new tests (excluding integration tests)? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew lgtm` (style, typechecking and tests) locally?

-----

- [x] AI was used to generate or assist with generating this PR.

AI (Claude) assisted with the diagnosis and drafting under human direction and review: locating the root cause and the introducing commit (`44f058edb5`), writing the code change, and updating the test. Verification so far: `brew style` passes locally on both changed files; the change was checked against the source — `download_name` feeds only `symlink_location`, while the cached blob is named from `resolved_basename` (the URL) and is therefore unaffected — and against the `Cask::Download` construction path. `brew lgtm` was subsequently run locally against this branch and passes: `brew typecheck` reports no errors; `brew tests --changed` runs the changed cask-download spec (all examples pass); and `brew style` finds no offenses on either changed file.

-----

### CODE REVIEW

File: Library/Homebrew/cask/download.rb [hunk 1]
```diff
@@ -163,22 +163,17 @@ def cache
...
-      temp_downloader = download_strategy.new(url.to_s, url_basename, version, mirrors: [], cache:, **url.specs)
-      return url_basename unless temp_downloader.is_a?(AbstractFileDownloadStrategy)
-
-      potential_symlink_length = temp_downloader.symlink_location.basename.to_s.length
-      max_filesystem_symlink_length = 255
-
-      return url_basename if potential_symlink_length < max_filesystem_symlink_length
-
-      cask.full_token.gsub("/", "--")
+      cask.token
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/download.rb:176]: This seems like a pretty big change. How many of homebrew/cask's casks have you tested this with?

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks! Good idea but worried that this will go 💥 after merge and/or invalidate caches incorrectly.

**Comment by aholland** [comment]
*Point of order up front: this reply was drafted collaboratively by me and Claude.*

Thanks — your questions prompted me to do some further testing. I wrote a harness to exercise the change against every cask and I've made it available here so you can run it yourself. I've also included a section below on how the bug arose — a history I very much had to get clear in my own mind before creating this PR.

**Short version:** I tested this against **all 7,739 casks** in homebrew/cask, offline, running the real production code and then the same code with this PR's fix applied. It fixes stale-download pruning for **7,224 / 7,224** file-download casks (the other 515 are git/VCS-backed and have no cache symlink, so they're not affected either way). The sweep also caught exactly the kind of post-merge breakage you were worried about — one cask whose *current* (not superseded) download would be wrongly deleted. That turned out to be a *separate*, pre-existing bug this fix merely unmasks: `stale_cask?` matches the version with a regular expression that mis-fires when a cask's version ends in punctuation. I've fixed it in the same PR with a one-line change to `cleanup.rb`, plus a regression test. And caches are **not** invalidated: the change only renames the cache *symlink*, never the content-addressed cached file.

Details below.

## How the bug arose

It's a refactor regression that three commits conspired to create:

- **2018 (`78658e4302`)** — cask cleanup is introduced. `Cleanup#stale_cask?` already resolves a cached cask by taking the text before the first `--` in the symlink name and loading it as a **cask token**. At the time, cask symlinks were token-named, so this worked fine.
- **2023 (`44f058edb5`, "Refactor formula, cask and Ruby source downloads to use shared code")** — casks are unified onto the shared `Downloadable` code, whose `download_name` is the **URL basename** (so the cached *blob* keeps its real download filename). This is the actual regression: casks now get a symlink-naming scheme that no longer matches what the 2018 cleanup code assumes. This same commit does touch `CaskLoader.load` in `cleanup.rb`, but doesn't reconcile the mismatch it was creating.
- **2025 (`b1bd71c412`, "Override download_name in Cask::Download to prevent long symlink names")** — the token gets retrofitted, but **only as a fallback for names that would exceed the 255-char filesystem limit**. That quietly fixed cleanup only for the rare cask whose name would otherwise be too long — just one in today's catalogue (`mimecast`) — while leaving the common short-name case, the vast majority, still broken.

So it was never a deliberate change to how cleanup resolves cached casks; the token assumption baked into cleanup in 2018 simply predates, and outlived, the 2023 refactor that broke it. This PR makes `Cask::Download#download_name` unconditionally return `cask.token`, reconciling all three.

## What I tested, and the result

I wrote a script that, for every cask and **with no network / no downloads**, names a *current* and a *superseded* cache download exactly the way the real, currently-loaded `download_name` names them, then runs the real `Cleanup#cleanup_cask` and records whether the superseded one gets pruned. Crucially it runs the genuine pre-fix code (checked out of git), not a re-implementation of it.

Running the identical sweep before and after the change:

| | superseded download pruned |
|---|---|
| **Before** (pre-fix `download.rb`) | **9 / 7,224** |
| **After** (this PR) | **7,224 / 7,224** |

The 9 that already worked pre-fix break down as 8 casks where the URL basename already equals the token, plus the single case the 2025 length-guard covered (`mimecast`, whose URL-basename name exceeds the filesystem limit and so fell back to the token). Everything else — 7,215 casks — silently leaked superseded downloads until the generic max-age sweep eventually caught them.

## The 💥 you predicted — found and fixed

Sweeping the whole catalogue surfaced exactly one genuine post-merge breakage, which I don't think would have been obvious by inspection: **`free-gpgmail`**. Its version is assembled from `version.csv` and ends in a trailing comma (`7.2,2023.3,`). Once the fix makes its download token-named, it becomes visible to `stale_cask?` for the first time — and `stale_cask?` anchored its version match with `\b`:

```ruby
basename.to_s.match?(/\A#{Regexp.escape(name)}--#{Regexp.escape(cask.version)}\b/)
```

A trailing comma followed by the extension's `.` is non-word → non-word, so `\b` can't match, the current version's download is judged stale, and it gets deleted on every `brew cleanup`. That's a real, if narrow, regression the naming change would otherwise have introduced — so I've included a one-line fix in the same PR (`\b` → `(?:\.|\z)`, i.e. "version followed by the extension dot or end of string"), which keeps the current download and still correctly treats `1.23` as stale when the version is `1.2`. It's covered by a new regression test (red/green verified) with a `local-comma-version` fixture.

## Cache invalidation: it doesn't

This is the part I was most careful about, and the answer is that nothing already cached is invalidated, because the change is decoupled from the cached file:

- **Installed software is untouched** — it lives in the Caskroom, not the cache.
- **Reinstalls still hit the cache.** The cached blob is content-addressed via a SHA256 of the URL (`cached_location` / `resolved_basename`), and install/reinstall finds it by that hash — *not* by the friendly symlink. `download_name` only affects the symlink, so no already-cached download is renamed or missed.
- **Old-style symlinks are left as-is.** They keep being skipped by `stale_cask?` exactly as before and age out within the normal max-age window — no migration, no mass deletion. The currently-installed version's download is never wrongly removed.
- Transitionally a cask may have both an old-style and a new-style symlink pointing at different version blobs — harmless, since blobs are content-addressed regardless of which symlink points at them (no duplication).

The scenario that *would* invalidate caches — renaming the cached blob itself — is explicitly what this change avoids: it renames only the symlink.

## Reproduce it yourself (offline, ~30s, nothing downloaded)

The script needs only `brew` — it enumerates every cask token from the JSON API (one cached metadata fetch, no cask blobs, no local homebrew/cask tap clone), and it redirects all cache operations into a throwaway temp dir, so **it never touches your real `~/Library/Caches`**. It runs fine on a bare install or a fresh container.

Save the script below, then on this branch:

```
./bin/brew ruby -- reproduce_cask_cleanup.rb --all      # -> 7224/7224 pruned
```

and to see the "before", run the same thing with `origin/main`'s `download.rb` in place (or just diff the `SUPERSEDED DOWNLOAD PRUNED` line between the two). `reproduce_cask_cleanup.rb <token>...` prints the per-cask mechanism for a specific cask.

<details>
<summary><code>reproduce_cask_cleanup.rb</code></summary>

```ruby
# Reproduce the `cask/download: fix cleanup` behaviour (Homebrew/brew#22911)
# against real casks, with no network access.
#
# This script does NOT model the old naming — it calls the real, production
# `Cask::Download#download_name` that is currently loaded. Run it once with the
# pre-fix `download.rb` checked out and once with the fix, and compare: the
# driver `run_reproduction.sh` does exactly that.
#
#   ./bin/brew ruby -- reproduce_cask_cleanup.rb              # curated sample
#   ./bin/brew ruby -- reproduce_cask_cleanup.rb firefox slack
#   ./bin/brew ruby -- reproduce_cask_cleanup.rb --all        # whole catalogue
#
# For each cask, purely from cache filenames (nothing is downloaded):
#   * name the current download the way the loaded code names it
#     (`downloader.symlink_location`), and a superseded download with the same
#     prefix (`Cask::Download#download_name`) but an old version;
#   * run the real `Cleanup#cleanup_cask`;
#   * record whether the superseded download was pruned and the current kept.
# Under the pre-fix code the prefix is the URL basename, which cleanup cannot
# resolve to a cask, so superseded downloads are NOT pruned (the bug). Under the
# fix the prefix is the cask token, so they are.

require "cask/cask_loader"
require "cask/download"
require "cleanup"
require "api"
require "fileutils"
require "stringio"
require "tmpdir"

# SAFETY: never touch the real Homebrew cache. Redirect every cache lookup this
# script makes (symlink naming AND cleanup pruning) into a throwaway temp dir, so
# it only ever creates and deletes files there. Your real ~/Library/Caches is
# untouched. The temp dir is removed at exit.
ISOLATED_CACHE = Pathname.new(Dir.mktmpdir("repro-cask-cleanup-"))
(ISOLATED_CACHE/"Cask").mkpath
Cask::Cache.singleton_class.prepend(Module.new do
  define_method(:path) { ISOLATED_CACHE/"Cask" }
end)
at_exit { FileUtils.remove_entry(ISOLATED_CACHE) if ISOLATED_CACHE.exist? }

CURATED = %w[
  transmission
  docker-desktop
  iplay
  emclient
  mimecast
  konica-minolta-bizhub-c759-c658-c368-c287-c3851-driver
].freeze

OLD = "0.0.0-superseded"

# The list of every cask token. By default this comes from Homebrew's JSON API
# (one cached metadata fetch — no cask blobs, no local homebrew/cask tap clone
# required), so this runs on a bare `brew` install, e.g. a fresh container.
# Set HOMEBREW_REPRO_CASK_DIR to a Casks directory to enumerate local .rb files
# instead (fully offline, or to pin a specific tap checkout).
def all_tokens
  if (dir = ENV["HOMEBREW_REPRO_CASK_DIR"])
    Dir.glob("#{dir}/**/*.rb").map { |p| File.basename(p, ".rb") }.sort
  else
    Homebrew::API::Cask.all_casks.keys.sort
  end
end

cleanup = Homebrew::Cleanup.new(cache: ISOLATED_CACHE)
Cask::Cache.path.mkpath

def run_cleanup(cleanup, cask, quiet)
  return cleanup.cleanup_cask(cask) unless quiet

  orig = $stdout
  $stdout = StringIO.new
  begin
    cleanup.cleanup_cask(cask)
  ensure
    $stdout = orig
  end
end

def check(cleanup, token, quiet:)
  current = superseded = nil
  cask = Cask::CaskLoader.load(token)
  dl = Cask::Download.new(cask)
  downloader = dl.downloader

  # Git/VCS-backed casks (many fonts) have no downloaded file or cache symlink.
  return { token:, kind: :vcs } unless downloader.is_a?(AbstractFileDownloadStrategy)

  current = downloader.symlink_location
  ext = current.extname
  prefix = dl.send(:download_name)          # the REAL production naming (old or new)
  superseded = Cask::Cache.path/"#{prefix}--#{OLD}#{ext}"

  FileUtils.mkpath current.dirname
  FileUtils.touch current

  superseded_unwritable = false
  unless superseded == current
    begin
      FileUtils.touch superseded
    rescue Errno::ENAMETOOLONG
      # The pre-fix URL-basename name can exceed the filesystem's 255-char limit.
      superseded_unwritable = true
    end
  end

  run_cleanup(cleanup, cask, quiet)

  {
    token:,
    symlink:               current.basename.to_s,
    prefix:,
    token_named:           prefix == cask.token,
    current_kept:          current.exist?,
    superseded_pruned:     !superseded_unwritable && !superseded.exist?,
    superseded_unwritable:,
  }
ensure
  [current, superseded].each { |f| f&.delete if f&.exist? }
end

if ARGV.delete("--all")
  tokens = all_tokens
  checked = token_named = kept = pruned = unwritable = vcs = 0
  errors = []
  not_pruned = []

  tokens.each do |token|
    begin
      r = check(cleanup, token, quiet: true)
    rescue => e
      errors << "#{token}: #{e.class}: #{e.message}"
      next
    end
    if r[:kind] == :vcs
      vcs += 1
      next
    end
    checked += 1
    token_named += 1 if r[:token_named]
    kept += 1 if r[:current_kept]
    pruned += 1 if r[:superseded_pruned]
    unwritable += 1 if r[:superseded_unwritable]
    not_pruned << r unless r[:superseded_pruned]
  end

  puts "file-download casks checked . #{checked}"
  puts "git/VCS casks (n/a) ........ #{vcs}"
  puts "load/compute errors ........ #{errors.size}"
  puts
  puts "symlink named by cask token  #{token_named}/#{checked}"
  puts "current download kept ...... #{kept}/#{checked}"
  puts "SUPERSEDED DOWNLOAD PRUNED . #{pruned}/#{checked}    <-- the fix's effect"
  puts "  (of the un-pruned, #{unwritable} had an old name too long to even store)"
  unless errors.empty?
    puts
    puts "Errors (#{errors.size}, first 20):"
    errors.first(20).each { |e| puts "  #{e}" }
  end
else
  tokens = ARGV.empty? ? CURATED : ARGV
  tokens.each do |token|
    r = check(cleanup, token, quiet: false)
    if r[:kind] == :vcs
      puts "SKIP  #{token} (git/VCS download — no cache symlink)"
      next
    end
    puts "#{token}"
    puts "        current symlink ........... #{r[:symlink]}"
    puts "        named by cask token ....... #{r[:token_named]}   (prefix=#{r[:prefix]})"
    puts "        current download kept ..... #{r[:current_kept]}"
    if r[:superseded_unwritable]
      puts "        superseded pruned ......... n/a (old-scheme name too long to store)"
    else
      puts "        superseded pruned ......... #{r[:superseded_pruned]}"
    end
  rescue => e
    puts "ERROR #{token}: #{e.class}: #{e.message}"
  end
end
```

</details>

File: Library/Homebrew/cask/download.rb (hunk 1 above)
Comment (aholland) [reply Library/Homebrew/cask/download.rb:176]: All of them now — see https://github.com/Homebrew/brew/pull/22911#issuecomment-4885926407

**Comment by MikeMcQuaid** [comment]
> Point of order up front: this reply was drafted collaboratively by me and Claude.

@aholland Please don't do that. I'm not going to read it. Please rewrite it in your own words or close this PR. The output from your AI is for you to read, not me. I need to feel like I'm talking to a human or there's little point in me not just prompting myself. One more chance.

**Comment by aholland** [comment]
Mike, I've put 7 hours into this this evening. I've read that response at least 8 times. I stand by my very carefully prepared words. The point of order was purely a politeness on account of your having asked me to clarify the same last time. The cache is not clearing for casks, and has not since 2023. In my case this fix cleared about 40GB.

**Comment by MikeMcQuaid** [comment]
@aholland My point is: I don't think they are your carefully prepared words, I think they are Claude's. I don't feel like I'm talking to a human here bar your last response. The writing is excessively verbose and poorly formatted in a way that makes it hard to read. I made an inline comment you've :+1:d but not responded to. I would rather prompt my own AI rather than try to parse this, sorry.

If you've genuinely spent 7 hours full time doing nothing else but staring at this PR: I'm sorry. Next time: spent less time and focus on writing as few words as possible required to communicate the information I've asked for. Read the above output from your LLM and summarise it.

**Comment by aholland** [comment]
Well @MikeMcQuaid if it's too wordy and formal, I'll give a shorter and more fluid version a go. It took me a significant part of last Tuesday to understand and fix this bug. This afternoon (I'm on New Zealand time BTW), seeing your review, I resolved to test it against every single Homebrew cask and did so. It was worth it because because it uncovered an edge case; not a flaw in the fix, but a second pre-existing bug that was masked until the fix. One of the 7724 casks, `free-gpgmail`, actually has a version string that ends with punctuation, which is allowed, but is something the regex in the `cleanup.rb` file tripped on. The method in question is  `stale_cask`, a private class method within Homebrew::Cleanup, line 222, edit on line 233. That regex uses a `\b`, word boundary, but in the fix uses `(?:\.|\z)` which won't trip up on the comma in the version string `7.2,2023.3,` which is the version string of `free-gpgmail`. Remember that all it's doing is checking for symlinks which match the cask token but where the version string does _not_ match the current version string.

While this change is safe, I agree that it is significant and as such imagined that you might want not just my word that I had checked every cask, but to be able to reproduce that yourself. That's why I attached the script, which is designed so that it can run in a Docker container, ie in a way that it safe for you, should you want to run these checks yourself.

Again, the casks are currently not clearing. The history that introduced the bug is evident in the relevant commits which are `78658e4302` (2018, no bug but introduced the mechanism, worked correctly, and used the tokens as does this fix), `44f058edb5` (2023, introduced the bug as part of a worthy effort to unify how downloads worked, so a DRY effort that went a bit wrong), and `b1bd71c412`, 2025 (Conditionally reintroduced use of the tokens in edge cases in which the URL base-name was too long for the file system. Out of the 7224 casks, one meets this criterion, `mimecast`). The bug also does not manifest for cases in which the URL base-name happens to match the token, which is 8 out of the 7224. So the cache is only clearing for 8+1 = 9 out of 7224 currently.  You may object that there are actually 7724+515 casks, but the extra 515 are not file-backed. They use Git and aren't part of this mechanism. Of course this affects all other taps too.

File: Library/Homebrew/cask/download.rb (hunk 1 above)
Comment (aholland) [reply Library/Homebrew/cask/download.rb:176]: @MikeMcQuaid commenting a second time on this because you said I didn't before. Is it okay to refer you to the main thread? The answer is that I have tested it againt all applicable Homebrew casks, which is 7739 less the 515 Git-backed ones, which equals 7724 casks. The fix works perfectly on all 7724, though 9 of them were edge cases that worked anyway.

Comment (aholland) [reply Library/Homebrew/cask/download.rb:176]: PS this explains your complaint about my not responding — I did not notice the pending badge, nor find this button way up at the top, until now. 

Image

Image

Comment (aholland) [reply Library/Homebrew/cask/download.rb:176]: PS would you have preferred if I had just put the entire response here? My reasoning was that it didn't apply only to this file, but you let me know.

**Comment by MikeMcQuaid** [comment]
@aholland Thanks for the additional responses here. After my last response I decided to work on my own PR instead, see https://github.com/Homebrew/brew/pull/22962



> The history that introduced the bug is evident in the relevant commits which are `78658e4302`

Note for future if you don't escape this e.g. 78658e4302 it provides a clickable link.
